test/rnn_seq_api.hpp

```diff
@@ -0,0 +1,1663 @@
+        auto& hLens  = xHiddenState.desc.GetLengths();
+
+        std::cout << " --batch_size " << inLens[0] << " --seq_len " << inLens[1] << " --in_vec "
+                  << inLens[2] << " --hid_h " << hLens[2] << " --num_layer " << rnnDesc.nLayers
+                  << " -r " << rnnDesc.dirMode << " -b " << rnnDesc.biasMode << " -p "
+                  << rnnDesc.inputMode << " -a " << rnnDesc.algoMode;
+
+        bool useDropout = !miopen::float_equal(miopen::deref(rnnDesc.dropoutDesc).dropout, 0);
+
+        std::cout << " --io_layout "
+                  << miopen::RNNDescriptor::getBaseLayoutFromDataTensor(input.desc)
+                  << " --use_dropout " << useDropout;
+        if(useDropout)
+            std::cout << " --dropout " << miopen::deref(rnnDesc.dropoutDesc).dropout;
+
+        auto& samplesLen = input.desc.GetSequenceLengthsVector();
+        std::cout << " --seq_len_array ";
+        for(int i = 0; i < inLens[0]; i++)
+        {
+            if(i < inLens[0] - 1)
+            {
+                std::cout << samplesLen.at(i) << ",";
+            }
+            else
+            {
+                std::cout << samplesLen.at(i);
+            }
+        }
+        std::cout << std::endl;
+
+        if(badtensor >= 0)
+        {
+            if(badtensor < 3)
+            {
+                std::cout << "FWD Train LSTM: " << std::endl;
+                switch(badtensor)
+                {
+                case(0):
+                    std::cout << "Output tensor output failed verification." << std::endl;
+                    break;
+                case(1):
+                    std::cout << "Hidden state tensor failed verification." << std::endl;
+                    break;
+                case(2): std::cout << "Cell state tensor failed verification." << std::endl; break;
+                default: break;
+                }
+            }
+            else if(badtensor < 6)
+            {
+                std::cout << "BWD Train LSTM: " << std::endl;
+                switch(badtensor)
+                {
+                case(3):
+                    std::cout << "Output tensor output failed verification." << std::endl;
+                    break;
+                case(4):
+                    std::cout << "Hidden state tensor failed verification." << std::endl;
+                    break;
+                case(5): std::cout << "Cell state tensor failed verification." << std::endl; break;
+                default: break;
+                }
+            }
+            else if(badtensor == 6)
+            {
+                std::cout << "WRW Train LSTM " << std::endl;
+            }
+        }
+    }
+};
+
+//****************************************************
+// RNN TRAIN
+//****************************************************
+
+template <class T>
+struct rnn_ref
+{
+    struct FWDResObj
+    {
+        std::vector<T> y;
+        std::vector<T> hy;
+        std::vector<T> cy;
+        FWDResObj(std::vector<T> fwd_y,
```

Comment:
[Q] Does ResObj stand for "result object"? Can we try to find better name? Basically, an instance of every class is an object, so this "Obj" looks redundant in the class name. But maybe it means something special?

Reply:
Yes, of course you can, I'm out of ideas.
This is just a set of local classes whose task is to make verification objects more explicit and simplify interaction with them.

Reply:
ResData, ResBuffers

Reply:
[Recommendation] Maybe just "FwdResult" or "FwdData""? So the fully qualified name would look like `rnn_ref::FwdResult` or `rnn_ref::FwdData` which seems clear enough. Up to you.

[resolved]